### PR TITLE
PERF: Improve topic serialization performance

### PR DESCRIPTION
**Pull request by angusmcleod**
@pmusaraj This is the first of a few performance related PRs. This one is focused on topics.

The next one will likely be focused on categories, and will come after https://github.com/discourse/discourse/pull/23969 is merged. I'm going to circle back to topics again after this is and the first categories performance PR is handled to see what additional gains can be made.

Note that all the actual performance gains here are being made in the topic `show` action. The AP plugin seems to have little discernible impact on topic lists (category lists are a different story) and I've added a spec to both (partially) demonstrate that and also as a guard against regression.

With respect to topic `show` I'll let the specs speak for themselves, but just to illustrate the gain being made there (and as a sanity check) these are the different profiles of a topic on my local. The topic has:
- Full Topic publication type (when AP is enabled)
- 7 posts from 4 different actors (users)

##### No AP plugin (vanilla discourse)
Screenshot 2023-10-19 at 14 16 28

##### AP plugin on this branch
Screenshot 2023-10-19 at 14 12 32

##### AP plugin on main
Screenshot 2023-10-19 at 14 14 21

### CODE REVIEW

File: plugin.rb [hunk 1]
```diff
@@ -509,10 +513,12 @@
...
   end
-  activity_pub_post_custom_fields.each do |field_name|
-    add_to_serializer(:post, "activity_pub_#{field_name}".to_sym) do
-      object.send("activity_pub_#{field_name}")
-    end
+  activity_pub_post_custom_field_names.each do |field_name|
+    add_to_serializer(
+      :post,
+      field_name.to_sym,
+      include_condition: -> { object.activity_pub_enabled }
```
Comment (pmusaraj) [inline plugin.rb:520]: This is nice. IIUC, this cleans up the JSON output of topics where AP isn't enabled.

File: spec/requests/topics_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,107 @@
+# frozen_string_literal: true
+
+RSpec.describe TopicsController do
+  ADDITIONAL_QUERY_LIMIT = 6
```
Comment (pmusaraj) [inline spec/requests/topics_controller_spec.rb:4]: Do the additional queries add up to 6 or is this set to 6 so that there's some extra room? If this number is precise and consistent across test runs, I'd prefer using that in tests so that an extra query can be more quickly detected when adding other features or making unrelated refactors to the plugin.

**Review by pmusaraj — APPROVED** [review]
Looks generally good to me, thanks for adding the extra specs for lists.

File: plugin.rb (hunk 1 above)
Comment (angusmcleod) [reply plugin.rb:520]: Yup, that's right, it also guards against unnecessary work (queries) when the plugin is disabled.

File: spec/requests/topics_controller_spec.rb (hunk 2 above)
Comment (angusmcleod) [reply spec/requests/topics_controller_spec.rb:4]: Yeah fair. I think they actually add up to 5 for first post and 6 for full topic. I'll revisit this specific point when I take a second look at topic performance (after categories).
